### PR TITLE
Add `dev_getBlockStats` RPC

### PR DESCRIPTION
**Pull request by athei**
This adds a new RPC:

```rust
pub struct BlockStats {
    pub witness_len: u64,
    pub witness_compact_len: u64,
    pub block_len: u64,
    pub block_num_extrinsics: u64,
}

#[rpc(name = "dev_getBlockStats")]
fn block_stats(&self, hash: Hash) -> Result<Option<BlockStats>>;
```

It is meant as a way to gain insides about the bottle necks of a given chain. While some information can only be learned through this RPC (witness related information) some of them are redundant: You can query the whole block and then measure the size of the block and the number of extrinsics. **However**, I think it still makes sense to include them as this saves us from transmitting the whole block to the RPC client (possibly over the internet) in order to learn about these statistics.

I have written a CLI tool that queries this draft version of the RPC:

```bash
cargo install --git https://github.com/athei/povstats.git --locked

# This will just connect to a locally running node and calls the RPC on each new block
povstats
```

The output when running this against a `substrate --dev` node without submitting any extrinsics to it looks like this:
```
27: Some(BlockStats { witness_len: 8179, witness_compact_len: 6124, witness_compressed_len: 5685, block_len: 226, block_num_extrinsics: 1 })
28: Some(BlockStats { witness_len: 7960, witness_compact_len: 5869, witness_compressed_len: 5412, block_len: 129, block_num_extrinsics: 1 })
29: Some(BlockStats { witness_len: 7894, witness_compact_len: 5901, witness_compressed_len: 5473, block_len: 129, block_num_extrinsics: 1 })
30: Some(BlockStats { witness_len: 8564, witness_compact_len: 6379, witness_compressed_len: 5918, block_len: 226, block_num_extrinsics: 1 })
31: Some(BlockStats { witness_len: 8535, witness_compact_len: 6382, witness_compressed_len: 5913, block_len: 226, block_num_extrinsics: 1 })
32: Some(BlockStats { witness_len: 8698, witness_compact_len: 6513, witness_compressed_len: 6065, block_len: 226, block_num_extrinsics: 1 })
33: Some(BlockStats { witness_len: 7842, witness_compact_len: 5913, witness_compressed_len: 5471, block_len: 129, block_num_extrinsics: 1 })
34: Some(BlockStats { witness_len: 7886, witness_compact_len: 5925, witness_compressed_len: 5473, block_len: 129, block_num_extrinsics: 1 })
35: Some(BlockStats { witness_len: 8520, witness_compact_len: 6497, witness_compressed_len: 6064, block_len: 226, block_num_extrinsics: 1 })
```

### CODE REVIEW

**Comment by kianenigma** [comment]
> It happens inside: sp_api::Core::execute_block which is called by the new RPC.

I am not sure with the abstraction level that you are using, and I don't see any code that specifies on top of which state you are executing the given block. What I can say for sure is that to execute any arbitrary block, you certainly need its parent state, which may not be available. This error that you said also rings a bell for me in `try-runtime`, and I recall the reason was not using the proper parent state.

**Comment by athei** [comment]
I thought I did correctly specify the parent state by calling the runtime API at the block's parent's hash:

```rust
runtime_api.execute_block(&BlockId::Hash(*parent_hash), block)
```

**Comment by kianenigma** [comment]
> I thought I did correctly specify the parent state by calling the runtime API at the block's parent's hash:
> 
> ```rust
> runtime_api.execute_block(&BlockId::Hash(*parent_hash), block)
> ```

that sounds correct to me. Perhaps to double check: if you pass an old-block's hash to a non-archive node, do you get an error?

File: client/rpc/src/chain/chain_full.rs
Comment (bkchr) [inline client/rpc/src/chain/chain_full.rs:87]: Your problem is that the block header contains the seal of the consensus engine. The runtime is "dumb" it just compares that the digests of the header that is the result of `execute_block` matches the header of the block you gave as input. The seal is added by the consensus engine after the runtime and is normally stripped before we call into the runtime.

**Review by bkchr — CHANGES_REQUESTED** [review]
In generally I'm against adding new rpc methods that are added for everybody by default. However, the functionality here could maybe added to the `state_traceBlock` rpc method. Or if you want to keep it separate, it should be put into its own crate.

File: client/rpc/src/chain/chain_full.rs
Comment (athei) [reply client/rpc/src/chain/chain_full.rs:87]: Is there anything I can do here? Should I just drop the last `Digest` using `block.header().digest_mut()`?

**Comment by athei** [comment]
> In generally I'm against adding new rpc methods that are added for everybody by default. However, the functionality here could maybe added to the state_traceBlock rpc method.

Isn't modifying existing RPCs that are included for anyone even worse? I can move it to its own crate so the RPC is optional.

File: client/rpc/src/chain/chain_full.rs
Comment (bkchr) [reply client/rpc/src/chain/chain_full.rs:87]: You could drop all of them that are `Digest::Seal`

**Comment by bkchr** [comment]
> Isn't modifying existing RPCs that are included for anyone even worse?

I mean we would just add some data to the output that you can just ignore on the calling site.

File: client/rpc/src/chain/chain_full.rs
Comment (athei) [reply client/rpc/src/chain/chain_full.rs:87]: Thanks a lot. I removed all of those before passing them into the runtime and it works now.

**Comment by athei** [comment]
> > Isn't modifying existing RPCs that are included for anyone even worse?
> 
> I mean we would just add some data to the output that you can just ignore on the calling site.

This is basically modifying the RPC in a backwards compatible way. So is adding a new RPC. Why is adding a new one worse?

**Comment by dvdplm** [comment]
My two cents:
- We don't have a good process to decide when/how/who to add RPCs (same for deprecation).
- Over time I've seen people think that "unsafe" RPCs are a way to signal that an RPC is experimental/debug-only/unstable. This is not the case, or rather, it's a bit of a semantic acrobatics. Unsafe means that an RPC should not be exposed to untrusted input, either because it does dangerous things or because it puts a lot of strain on the node and thus exposing a DoS veector.
- @tomaka 's approach when re-designing the RPCs is to be very explicit about which RPCs are stable and which aren't, by including "unstable" in the name. I think that's one way to go about it (preferably letting the proc macro do the heavy lifting).
- Given that up until now, adding or changing an RPC has been pretty much "do whatever you want", I think it's unfair to block this PR and I don't think the functionality fits very well as an extension to `traceBlock`


In other words: no objections to merge this from me (but it'd be good to start making up our minds about how to add RPCs, the sooner the better).

**Comment by bkchr** [comment]
> In other words: no objections to merge this from me (but it'd be good to start making up our minds about how to add RPCs, the sooner the better).

This is relative easy. Don't add RPCs automatically to every node building on Substrate. Adding RPC functionality to Substrate isn't bad and we can name them `unstable` or whatever. However, people building on Substrate should decide on their own what kind of RPCs they want to expose.

File: client/rpc/src/chain/chain_full.rs
Comment (athei) [reply client/rpc/src/chain/chain_full.rs:87]: Yes. PR is still WIP (draft). The unsafe check is missing.

File: client/rpc/src/chain/mod.rs
Comment (athei) [reply client/rpc/src/chain/mod.rs:78]: Yes. I will add documentation once there is an agreement where to put this RPC.

**Comment by athei** [comment]
> > In other words: no objections to merge this from me (but it'd be good to start making up our minds about how to add RPCs, the sooner the better).
> 
> This is relative easy. Don't add RPCs automatically to every node building on Substrate. Adding RPC functionality to Substrate isn't bad and we can name them `unstable` or whatever. However, people building on Substrate should decide on their own what kind of RPCs they want to expose.

So a new category within `sc-rpc` would be fine? Putting it into a new crate just adds more boilerplate for no benefit.

**Comment by bkchr** [comment]
Yeah, a new category in `sc-rpc` is okay :) I mean, I have no idea how the structure with jsonrpsee looks, that was the reason I proposed to speak with @dvdplm :)

**Comment by athei** [comment]
@dvdplm said: The less boilerplate the better for out transition. So even adding a new category has a lot of boilerplate and makes the transition more work. That said, we need to balance this against how bad it us for RPCs to "just show up" when you bump substrate.

**Comment by dvdplm** [comment]
> I mean, I have no idea how the structure with jsonrpsee looks, that was the reason I proposed to speak with @dvdplm :)

The amount of work for devs is quite similar, maybe slightly less but not by much.

**Comment by athei** [comment]
Okay so I will add a new category then. I still have no idea how to call this category. I was thinking about calling it `dev` to clearly indicate that it will contain some seriously wonky RPCs that are not really stable. We could then add new stuff there without much concern because wouldn't add this category to their production node anyways.

What do you think?

File: primitives/runtime/src/lib.rs
Comment (athei) [reply primitives/runtime/src/lib.rs]: Added

File: client/rpc/src/chain/chain_full.rs
Comment (athei) [reply client/rpc/src/chain/chain_full.rs]: Added the check.

**Comment by athei** [comment]
Ready for review now:
* Moved RPC to a new category.
* Added docs.
* Removed the `witness_compressed_len`. It is too assumptions. Client can compress itself if wanted.

File: client/rpc-api/src/dev/error.rs
Comment (niklasad1) [inline client/rpc-api/src/dev/error.rs:55]:
```suggestion
				message: e.to_string(),
```

File: client/rpc-api/src/dev/error.rs
Comment (niklasad1) [inline client/rpc-api/src/dev/error.rs:60]:
```suggestion
				message: e.to_string(),
```

File: client/rpc-api/src/dev/error.rs
Comment (niklasad1) [inline client/rpc-api/src/dev/error.rs:65]:
```suggestion
				message: e.to_string(),
```

File: primitives/runtime/src/lib.rs
Comment (bkchr) [inline primitives/runtime/src/lib.rs:923]: For a dev api. Do we really need to keep the traffic low? Sounds like some "over optimization"

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:75]:
```suggestion
				// Remove the `Seal` to ensure we have the number of digests as expected by the runtime.
				header.digest_mut().logs.retain(|item| !matches!(item, DigestItem::Seal(_, _)));
```

File: primitives/runtime/src/lib.rs
Comment (bkchr) [inline primitives/runtime/src/lib.rs:909]: And why is that defined in `sp-runtime`? Can that type not just be defined alongside the rpc api declaration?

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:26]:
```suggestion
use sc_client_api::{BlockBackend, HeaderBackend};
```

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:62]:
```suggestion
	Client: BlockBackend<Block> + HeaderBackend<Block> + ProvideRuntimeApi<Block> + Send + Sync + 'static,
```

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:92]:
```suggestion
		let parent_header = {
			let parent_hash = *block.header().parent_hash();
			let parent_header = self
				.client
				.block(BlockId::Hash(parent_hash))
				.map_err(|e| Error::BlockQueryError(Box::new(e)))?;
			if let Some(header) = parent_header {
				header
			} else {
				return Ok(None)
			}
		};
```

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:95]:
```suggestion
		let pre_root = *parent_header.state_root();
```

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:99]:
```suggestion
			.execute_block(&BlockId::Hash(parent_header.hash()), block)
```

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:108]:
```suggestion
			.encoded_size() as u64;
```

File: client/rpc/src/dev/mod.rs
Comment (bkchr) [inline client/rpc/src/dev/mod.rs:111]:
```suggestion
			witness_compact_len,
```

File: client/rpc-api/src/dev/error.rs
Comment (bkchr) [inline client/rpc-api/src/dev/error.rs:30]:
```suggestion
/// Dev RPC Result type.
pub type Result<T> = std::result::Result<T, Error>;

/// Dev RPC future Result type.
pub type FutureResult<T> = jsonrpc_core::BoxFuture<Result<T>>;

/// Dev RPC errors.
```

File: client/rpc-api/src/dev/error.rs
Comment (bkchr) [inline client/rpc-api/src/dev/error.rs:19]:
```suggestion
//! Error helpers for Dev RPC module.
```

File: client/rpc-api/src/dev/error.rs
Comment (bkchr) [inline client/rpc-api/src/dev/error.rs:3]:
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

File: client/rpc-api/src/dev/mod.rs
Comment (bkchr) [inline client/rpc-api/src/dev/mod.rs:3]:
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

File: client/rpc-api/src/dev/mod.rs
Comment (bkchr) [inline client/rpc-api/src/dev/mod.rs:31]: Why are we generic over `BlockStats`? As said for in the runtime crate, we should move the definition of the type here above the trait.

File: client/rpc-api/src/dev/mod.rs
Comment (bkchr) [inline client/rpc-api/src/dev/mod.rs:36]:
```suggestion
	/// This function will require the specified block and its parent to be available
	/// at the queried node. If either the specified block or the parent are not available, 
	/// this function will return `None`.
```

The parent is also required for having its state to re-execute the block on this state.

**Review by bkchr — APPROVED** [review]
Most of the stuff are just nitpicks, otherwise it looks good.

File: primitives/runtime/src/lib.rs
Comment (athei) [reply primitives/runtime/src/lib.rs:923]: I would rather argue that leaving it out would be over optimization: It is very easy to include this data. If we'd leave it out the client needs to refetch the block which makes the RPC client more complicated and brittle.

The reason for leaving it out would be to deduplicate. Deduplication is just another form of optimization :)

File: primitives/runtime/src/lib.rs
Comment (athei) [reply primitives/runtime/src/lib.rs:909]: I am totally lost where to define that. If you think it should be beside the RPC I can move it there. I just saw that no type was defined there and figured that this isn't the place.

Comment (bkchr) [reply primitives/runtime/src/lib.rs:909]: This type doesn't make any sense outside of this rpc. So, we should define it near to the rpc api.

File: primitives/runtime/src/lib.rs
Comment (athei) [reply primitives/runtime/src/lib.rs:923]: I can update the comment.

File: primitives/runtime/src/lib.rs
Comment (athei) [reply primitives/runtime/src/lib.rs]: Moved it.

File: client/rpc-api/src/dev/mod.rs
Comment (dvdplm) [inline client/rpc-api/src/dev/mod.rs:19]:
```suggestion
//! Substrate dev API containing RPCs that are mainly meant for debugging and stats collection for developers. The endpoints in this RPC module are not meant to be available to non-local users and are all marked `unsafe`.
```

File: client/rpc-api/src/dev/mod.rs
Comment (dvdplm) [inline client/rpc-api/src/dev/mod.rs:35]:
```suggestion
	/// The length in bytes of the storage proof after compaction.
```

(To me "after compression" sounds better, but then it should be `witness_compressed_len` too… Either is fine.)

File: client/rpc-api/src/dev/mod.rs
Comment (dvdplm) [inline client/rpc-api/src/dev/mod.rs:40]:
```suggestion
	/// This information can also be acquired by downloading the whole block. This merely
	/// saves some complexity on the client side.
```

File: client/rpc-api/src/dev/mod.rs
Comment (dvdplm) [inline client/rpc-api/src/dev/mod.rs:45]:
```suggestion
	/// This information can also be acquired by downloading the whole block. This merely
	/// saves some complexity on the client side.
```

File: client/rpc-api/src/dev/mod.rs
Comment (dvdplm) [inline client/rpc-api/src/dev/mod.rs:59]:
```suggestion
	/// This function requires the specified block and its parent to be available
	/// at the queried node. If either the specified block or the parent is pruned,
	/// this function will return `None`.
```

Is there any other reason than pruning that a block is unavailable?

File: client/rpc/src/dev/mod.rs
Comment (dvdplm) [inline client/rpc/src/dev/mod.rs:19]:
```suggestion
//! Implementation of the [`DevApi`] trait providing debug utilities for Substrate based blockchains.
```

**Review by dvdplm — APPROVED** [review]
Code LGTM. Left some nitpicks. :)

File: client/rpc-api/src/dev/mod.rs
Comment (athei) [reply client/rpc-api/src/dev/mod.rs:35]: We call it compact because there is compression, too. We compact the proof and then we compress it (with zstd). So using the word compression here would be confusing.

**Comment by athei** [comment]
bot merge

**Pull request merged**
